File: android-lib/src/androidTest/java/com/kinvey/androidTest/cache/CacheTest.java
package com.kinvey.androidTest.cache;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.RenamingDelegatingContext;
import android.test.suitebuilder.annotation.SmallTest;

import com.google.api.client.json.GenericJson;
import com.kinvey.android.Client;
import com.kinvey.android.cache.RealmCacheManager;
import com.kinvey.android.model.User;
import com.kinvey.java.cache.ICache;
import com.kinvey.java.cache.ICacheManager;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.*;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;
import java.util.List;

/**
 * Created by Prots on 1/27/16.
 */
@RunWith(AndroidJUnit4.class)
@SmallTest
public class CacheTest {
    ICacheManager cacheManager;
    @Before
    public void setup(){
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        cacheManager = new Client.Builder(mMockContext).build().getCacheManager();
        User user = new User();
        user.setId("testId");
        Client.sharedInstance().setActiveUser(user);
    }

    @Test
    public void testSaveSingleInsert(){
        ICache<SampleGsonObject1> cache = cacheManager.getCache("test", SampleGsonObject1.class, Long.MAX_VALUE);

        SampleGsonObject1 obj = new SampleGsonObject1();
        String id = UUID.randomUUID().toString();
        obj._id = id;
        obj.title = "test";
        assertEquals(cache.save(obj)._id, id);

        SampleGsonObject1 ret = cache.get(id);
        //test the same
        assertNotNull(ret);
        assertEquals(ret._id, obj._id);
        assertEquals(ret.title, obj.title);
    }

    @Test
    public void testSaveSingleUpdate(){
        ICache<SampleGsonObject1> cache = cacheManager.getCache("test", SampleGsonObject1.class, Long.MAX_VALUE);
        String id = "test";

        SampleGsonObject1 obj = new SampleGsonObject1();
        obj.put("_id", id);
        obj.put("title", "testTile");

        assertEquals(cache.save(obj)._id, id);

        obj.put("title", "testTitle2");
        assertEquals(cache.save(obj)._id, id);

        SampleGsonObject1 ret = cache.get(id);

        assertNotNull(ret);
        assertEquals(ret.title, "testTitle2");

    }

    @Test
    public void testMultipleInsert(){
        ICache<SampleGsonObject1> cache = cacheManager.getCache("test", SampleGsonObject1.class, Long.MAX_VALUE);
        String id = "test";

        List<SampleGsonObject1> items = new ArrayList<SampleGsonObject1>();

        for (int i = 0 ; i < 100; i++){
            items.add(new SampleGsonObject1(String.valueOf(i), "multipleInsert"+i));
        }

        List<SampleGsonObject1> saved = cache.save(items);

        assertNotNull(saved);

        assertEquals(100, saved.size());
        boolean idsOk = true;
        for (int i = 0 ; i < 100; i++){
            idsOk &= String.valueOf(i).equalsIgnoreCase(saved.get(i).get("_id").toString());
        }
        assertTrue("all ids are right and in right order", idsOk);

        List<String> ids = new ArrayList<String>();
        for (SampleGsonObject1 obj : saved){
            ids.add(obj._id.toString());
        }

        List<SampleGsonObject1> cachedObjects = cache.get(ids);

        assertEquals(100, cachedObjects.size());

    }


    @Test
    public void testDelete(){
        ICache<SampleGsonObject1> cache = cacheManager.getCache("test", SampleGsonObject1.class, Long.MAX_VALUE);

        List<SampleGsonObject1> items = new ArrayList<SampleGsonObject1>();

        List<String> ids = new ArrayList<String>();

        for (int i = 0 ; i < 100; i++){
            items.add(new SampleGsonObject1(String.valueOf(i), "deleteTest"+i));
            ids.add(String.valueOf(i));
        }

        List<SampleGsonObject1> saved = cache.save(items);

        assertNotNull(saved);

        cache.delete(ids);

        List<SampleGsonObject1> cachedObjects = cache.get(ids);

        assertEquals(0, cachedObjects.size());
    }

    @Test
    public void testSaveWithInner(){
        ICache<SampleGsonWithInner> cache = cacheManager.getCache("testInner2", SampleGsonWithInner.class, Long.MAX_VALUE);

        List<SampleGsonWithInner> items = new ArrayList<SampleGsonWithInner>();

        List<String> ids = new ArrayList<String>();

        for (int i = 0 ; i < 100; i++){
            items.add(new SampleGsonWithInner(String.valueOf(i), new SampleGsonObject1(String.valueOf(i), "test"+i)));
            ids.add(String.valueOf(i));
        }

        List<SampleGsonWithInner> saved = cache.save(items);

        assertNotNull(saved);


        List<SampleGsonWithInner> cachedObjects = cache.get(ids);

        assertEquals(100, cachedObjects.size());

        for (int i = 0 ; i < cachedObjects.size() ; i ++){
            SampleGsonWithInner res = cachedObjects.get(i);
            assertEquals(res.getDetails()._id, String.valueOf(i));
        }

    }

    @Test
    public void testSaveWithInnerList(){
        ICache<SampleGsonWithInnerList> cache = cacheManager.getCache("testInnerWithList3", SampleGsonWithInnerList.class, Long.MAX_VALUE);

        List<SampleGsonWithInnerList> items = new ArrayList<SampleGsonWithInnerList>();

        List<String> ids = new ArrayList<String>();

        for (int i = 0 ; i < 100; i++){
            List<SampleGsonObject1> details = new ArrayList<SampleGsonObject1>();
            details.add(new SampleGsonObject1(String.valueOf(i), "test"+i));
            items.add(new SampleGsonWithInnerList(String.valueOf(i), details));
            ids.add(String.valueOf(i));
        }

            List<SampleGsonWithInnerList> saved = cache.save(items);

        assertNotNull(saved);


        List<SampleGsonWithInnerList> cachedObjects = cache.get(ids);

        assertEquals(100, cachedObjects.size());

        for (int i = 0 ; i < cachedObjects.size() ; i ++){
            SampleGsonWithInnerList res = cachedObjects.get(i);
            assertEquals(res.getDetails().get(0)._id, String.valueOf(i));
        }

    }
}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/cache/CacheTest.kt
package com.kinvey.androidTest.store.data.cache

import android.content.Context
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.kinvey.android.Client.Builder
import com.kinvey.android.Client.Companion.sharedInstance
import com.kinvey.android.model.User
import com.kinvey.java.cache.ICache
import com.kinvey.java.cache.ICacheManager
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

/**
 * Created by Prots on 1/27/16.
 */
@RunWith(AndroidJUnit4::class)
@SmallTest
class CacheTest {

    var cacheManager: ICacheManager? = null

    @Before
    fun setup() {
        val mMockContext: Context? = InstrumentationRegistry.getInstrumentation().targetContext
        cacheManager = Builder<User>(mMockContext).build().cacheManager
        val user = User()
        user.id = "testId"
        sharedInstance().activeUser = user
    }

    @Test
    fun testSaveSingleInsert() {
        val cache = cacheManager?.getCache("test", SampleGsonObject1::class.java, Long.MAX_VALUE)
        val obj = SampleGsonObject1()
        val id = UUID.randomUUID().toString()
        obj.id = id
        obj.title = "test"
        assertEquals(cache?.save(obj)?.id, id)
        val ret = cache?.get(id)
        //test the same
        assertNotNull(ret)
        assertEquals(ret?.id, obj.id)
        assertEquals(ret?.title, obj.title)
    }

    @Test
    fun testSaveSingleUpdate() {
        val cache = cacheManager?.getCache("test", SampleGsonObject1::class.java, Long.MAX_VALUE)
        val id = "test"
        val obj = SampleGsonObject1()
        obj["_id"] = id
        obj["title"] = "testTile"
        assertEquals(cache?.save(obj)?.id, id)
        obj["title"] = "testTitle2"
        assertEquals(cache?.save(obj)?.id, id)
        val ret = cache?.get(id)
        assertNotNull(ret)
        assertEquals(ret?.title, "testTitle2")
    }

    @Test
    fun testMultipleInsert() {
        val cache = cacheManager?.getCache("test", SampleGsonObject1::class.java, Long.MAX_VALUE)
        val id = "test"
        val items: MutableList<SampleGsonObject1> = ArrayList()
        for (i in 0..99) {
            items.add(SampleGsonObject1(i.toString(), "multipleInsert$i"))
        }
        val saved = cache?.save(items) ?: listOf()
        assertNotNull(saved)
        assertEquals(100L, saved.size.toLong())
        var idsOk = true
        (0..99).forEach { i ->
            idsOk = idsOk and i.toString().equals(saved[i]["_id"].toString(), ignoreCase = true)
        }
        assertTrue("all ids are right and in right order", idsOk)
        val ids: MutableList<String> = ArrayList()
        for (obj in saved) {
            ids.add(obj.id.toString())
        }
        val cachedObjects = cache?.get(ids)
        assertEquals(100L, cachedObjects?.size?.toLong())
    }

    @Test
    fun testDelete() {
        val cache = cacheManager?.getCache("test", SampleGsonObject1::class.java, Long.MAX_VALUE)
        val items: MutableList<SampleGsonObject1> = ArrayList()
        val ids: MutableList<String> = ArrayList()
        (0..99).forEach { i ->
            items.add(SampleGsonObject1(i.toString(), "deleteTest$i"))
            ids.add(i.toString())
        }
        val saved = cache?.save(items)
        assertNotNull(saved)
        cache?.delete(ids)
        val cachedObjects = cache?.get(ids)
        assertEquals(0L, cachedObjects?.size?.toLong())
    }

    @Test
    fun testSaveWithInner() {
        val cache = cacheManager?.getCache("testInner2", SampleGsonWithInner::class.java, Long.MAX_VALUE)
        val items: MutableList<SampleGsonWithInner> = ArrayList()
        val ids: MutableList<String> = ArrayList()
        (0..99).forEach { i ->
            items.add(SampleGsonWithInner(i.toString(), SampleGsonObject1(i.toString(), "test$i")))
            ids.add(i.toString())
        }
        val saved = cache?.save(items)
        assertNotNull(saved)
        val cachedObjects = cache?.get(ids)
        assertEquals(100L, cachedObjects?.size?.toLong())
        cachedObjects?.indices?.forEach { i ->
            val res = cachedObjects[i]
            assertEquals(res.details?.id, i.toString())
        }
    }

    @Test
    fun testSaveWithInnerList() {
        val cache = cacheManager?.getCache("testInnerWithList3", SampleGsonWithInnerList::class.java, Long.MAX_VALUE)
        val items: MutableList<SampleGsonWithInnerList> = ArrayList()
        val ids: MutableList<String> = ArrayList()
        (0..99).forEach { i ->
            val details: MutableList<SampleGsonObject1> = ArrayList()
            details.add(SampleGsonObject1(i.toString(), "test$i"))
            items.add(SampleGsonWithInnerList(i.toString(), details))
            ids.add(i.toString())
        }
        val saved = cache?.save(items)
        assertNotNull(saved)
        val cachedObjects = cache?.get(ids)
        assertEquals(100L, cachedObjects?.size?.toLong())
        cachedObjects?.indices?.forEach { i ->
            val res = cachedObjects[i]
            assertEquals(res.details?.get(0)?.id, i.toString())
        }
    }

    @Test
    fun testGetTTL() {
        val ttl = 123456L
        assertEquals(ttl, cacheManager?.getCache("testInnerWithList3", SampleGsonWithInnerList::class.java, ttl)?.ttl)
    }
}